### PR TITLE
Branch comparison folding and order optimizations

### LINKED ISSUE #606: Optimize branches
Given input WebAssembly like this:

```wasm
  (block $label$1
   (if
    (i32.eqz
     (get_global $global$6)
    )
    (return
     (get_local $11)
    )
...
```

The Cranelift IR looks like this:

```clif
...
@000b   v10 = load.i32 notrap aligned v3+136
        v58 = ifcmp_imm v10, 0
@000e   brif ne v58, ebb3
@0012   jump ebb1(v8)

ebb3:
...
```

We should optimize this by swapping the conditional and unconditional branches, so that we have

```
         brif eq v58, ebb1(v8)
         jump ebb3
```
where the `jump` can then turn into a fallthrough. I think we can do this in `simple_preopt`, since it's unlikely the legalization will expose more opportunities to do this.

### PR DESCRIPTION
**Pull request by tyler**
Fixes #606.

This PR add two optimizations to `simple_preopt.rs`. The first is to fold code where we compare against zero, then use an explicit conditional branch. For instance:
```
v1 = icmp_imm v0, 0
brz v1, ebb1
```
becomes
```
brnz v0, ebb1
```
For x86-64 this doesn't actually make much difference, as it compiles to a `test` rather than a `cmp`, which should be very near equivalent. But it does save a byte on a very common operation, so hey.

The second is reordering branches at the end of ebbs to encourage fallthroughs. As pointed out in #606, it's common for wasm code to be compiled like this:
```
    brz v0, ebb1
    jump ebb2
ebb1:
    ...
ebb2:
    ...
```
Which ends up being compiled as a conditional jump followed by a jump. However, we can switch those instructions by flipping the destinations and inverting the conditional jump's condition, like so:
```
    brnz v0, ebb2
    fallthrough ebb1
ebb1:
    ...
ebb2:
    ...
```
And when we do that we eliminate the latter jump entirely.

### CODE REVIEW

File: lib/codegen/src/simple_preopt.rs [hunk 1]
```diff
@@ -536,11 +542,279 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+}
+
+fn branch_order(pos: &mut FuncCursor, cfg: &mut ControlFlowGraph, ebb: Ebb, inst: Inst) {
+    let info = match pos.func.dfg[inst] {
+        InstructionData::Jump {
+            opcode: Opcode::Jump,
+            destination,
+            ref args,
+        } => {
+            if let Some(next_ebb) = pos.func.layout.next_ebb(ebb) {
```
Comment (bjorn3) [inline lib/codegen/src/simple_preopt.rs]: Could you please do it like:

```rust
let next_ebb = if let Some(next_ebb) = pos.func.layout.next_ebb(ebb) {
    next_ebb
} else {
    return;
};

if destination == next_ebb {
    return;
}

// ...
```

Here and below to reduce the indentation significantly.

File: lib/codegen/src/simple_preopt.rs [hunk 2]
```diff
@@ -536,11 +542,279 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+                if destination == next_ebb {
+                    return;
+                }
+
+                if let Some(prev_inst) = pos.func.layout.prev_inst(inst) {
+                    let prev_inst_data = &pos.func.dfg[prev_inst];
+                    if !prev_inst_data.opcode().is_branch() {
```
Comment (sunfishcode) [inline lib/codegen/src/simple_preopt.rs]: This test is redundant, as `branch_destination()` below will just return None if prev_inst_data is not a branch.

File: lib/codegen/src/simple_preopt.rs [hunk 3]
```diff
@@ -536,11 +542,279 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+                            args: args.clone(),
+                            kind: BranchOptKind::NotEqualZero,
+                        },
+                        IntCC::Equal => BranchOptInfo {
+                            br_inst: inst,
+                            cmp_arg: cmp_arg,
+                            destination: destination,
+                            args: args.clone(),
+                            kind: BranchOptKind::EqualZero,
+                        },
```
Comment (sunfishcode) [inline lib/codegen/src/simple_preopt.rs]: The duplication between the NotEqual and Equal arms here can be factored out by doing `let kind = match br_cond { ... };` to compute just the kind, and then you can do a single `BranchOptInfo { ... }`.

File: lib/codegen/src/simple_preopt.rs [hunk 4]
```diff
@@ -536,11 +542,279 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+                                    Opcode::Brnz => BranchOrderInfo {
+                                        term_inst: inst,
+                                        term_inst_args: args.clone(),
+                                        term_dest: destination,
+                                        cond_inst: prev_inst,
+                                        cond_arg: cond_arg,
+                                        cond_inst_args: prev_args.clone(),
+                                        cond_dest: *cond_dest,
+                                        kind: BranchOrderKind::BrnzToBrz,
+                                    },
```
Comment (sunfishcode) [inline lib/codegen/src/simple_preopt.rs]: Ditto here. The kind is the only thing that's different between the two cases, so you can factor that out.

File: lib/codegen/src/simple_preopt.rs [hunk 5]
```diff
@@ -536,11 +542,279 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+    args: ValueList,
+    kind: BranchOptKind,
+}
+
+enum BranchOptKind {
+    EqualZero,
+    NotEqualZero,
+}
+
+fn branch_opt(pos: &mut FuncCursor, inst: Inst) {
```
Comment (sunfishcode) [inline lib/codegen/src/simple_preopt.rs]: Please add a brief comment here describing what this function does.

File: lib/codegen/src/simple_preopt.rs [hunk 6]
```diff
@@ -536,11 +542,279 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+}
+
+enum BranchOrderKind {
+    BrzToBrnz,
+    BrnzToBrz,
+    InvertIntCond(IntCC),
+    InvertFloatCond(FloatCC),
+}
+
+fn branch_order(pos: &mut FuncCursor, cfg: &mut ControlFlowGraph, ebb: Ebb, inst: Inst) {
```
Comment (sunfishcode) [inline lib/codegen/src/simple_preopt.rs]: This too :-).

File: lib/codegen/src/simple_preopt.rs [hunk 7]
```diff
@@ -536,11 +542,279 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+            destination,
+            ref args,
+        } => {
+            let first_arg = {
+                let args = pos.func.dfg.inst_args(inst);
+                args[0]
+            };
+            if let ValueDef::Result(iconst_inst, _) = pos.func.dfg.value_def(first_arg) {
+                if let InstructionData::BinaryImm {
+                    opcode: Opcode::IfcmpImm,
```
Comment (sunfishcode) [inline lib/codegen/src/simple_preopt.rs]: This is admittedly not document as clearly as it could be, but we currently don't expect frontends to produce flags-using instructions like `ifcmp` or `ifcmp_imm` themselves. Regular `icmp` is lowered to `ifcmp` by the legalizer on platforms that use a flags register on x86 and in the future other architectures. Since the code here is in simple_preopt, the legalizer hasn't run yet, so in theory we don't expect to see `ifcmp_imm` occur at this point. So one question here is: are you seeing `ifmp_imm` at this point?

If so, I'm not strongly attached to the "frontends shouldn't use flags" rule; we could just as easily have frontends produce flags-using code and have the legalizer lower it into other things on platforms that don't use flags, and it's not immediately obvious which way is better.

File: lib/codegen/src/simple_preopt.rs [hunk 8]
```diff
@@ -536,11 +542,279 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+            opcode: Opcode::Brif,
+            cond: br_cond,
+            destination,
+            ref args,
+        } => {
+            let first_arg = {
+                let args = pos.func.dfg.inst_args(inst);
+                args[0]
+            };
+            if let ValueDef::Result(iconst_inst, _) = pos.func.dfg.value_def(first_arg) {
```
Comment (bjorn3) [inline lib/codegen/src/simple_preopt.rs]: Please do this like the following to avoid that immense amount of indentation:

```rust
let iconst_inst = if let ValueDef::Result(iconst_inst, _) = pos.func.dfg.value_def(first_arg) {
    iconst_inst
} else {
    return;
};
```

File: lib/codegen/src/simple_preopt.rs (hunk 7 above)
Comment (tyler) [reply lib/codegen/src/simple_preopt.rs]: Error on my end. I've fixed this to match the branch ops that frontends are actually supposed to use. I was blindly going by the code example in #606 and hadn't tested it against real code until today. :)

File: Cargo.toml [hunk 9]
```diff
@@ -14,7 +14,7 @@ name = "clif-util"
 path = "src/clif-util.rs"
 
 [dependencies]
-cfg-if = "0.1"
+cfg-if = "0.1.6"
```
Comment (tyler) [inline Cargo.toml]: This appears to have been underspecified. With v0.1.4 locally, I was getting errors about missing macros. Presumably due to lack of this commit: https://github.com/alexcrichton/cfg-if/commit/67e01e5f0ea0e27d0bba38d3da7d73cda52d8be4

**Comment by tyler** [comment]
I think this is ready for another round of reviews.

**Comment by bnjbvr** [comment]
Hi @tyler ! Any idea when/if you'd have time to address the review comments and rebase this PR here? Or can we help you moving this forward by taking over the remaining changes?

**Comment by tyler** [comment]
hey @bnjbvr, yeah I can get back to this soon. I'll try to get to that this afternoon.

**Comment by tyler** [comment]
(If I don't, feel free to take it over. :))

**Comment by tyler** [comment]
@sunfishcode or @bnjbvr, do you want to take another look before merge? I think I addressed all the things.

File: cranelift-codegen/src/simple_preopt.rs [hunk 10]
```diff
@@ -536,11 +542,264 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+            }
+            BranchOptKind::NotEqualZero => {
+                let args = info.args.as_slice(&pos.func.dfg.value_lists)[1..].to_vec();
+                pos.func
+                    .dfg
+                    .replace(info.br_inst)
+                    .brnz(info.cmp_arg, info.destination, &args);
+            }
+        }
+    */
```
Comment (bjorn3) [inline cranelift-codegen/src/simple_preopt.rs]: Remove this?

**Comment by bnjbvr** [comment]
Thanks @tyler! I'll try to take a look at this early next week.

File: cranelift-codegen/src/simple_preopt.rs (hunk 10 above)
Comment (tyler) [reply cranelift-codegen/src/simple_preopt.rs]: I'm an idiot. Thanks. :)

File: cranelift-codegen/src/simple_preopt.rs [hunk 11]
```diff
@@ -536,11 +542,245 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+            }
+        }
+
+        _ => return,
+    };
+
+    let cond_args = {
+        info.cond_inst_args
+            .as_slice(&pos.func.dfg.value_lists)
+            .to_vec()
```
Comment (bnjbvr) [inline cranelift-codegen/src/simple_preopt.rs]: It seems all the uses of `cond_args` can use a slice; could we remove the vec heap allocation here?

File: cranelift-codegen/src/simple_preopt.rs [hunk 12]
```diff
@@ -536,11 +542,245 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+    };
+    let term_args = {
+        info.term_inst_args
+            .as_slice(&pos.func.dfg.value_lists)
+            .to_vec()
```
Comment (bnjbvr) [inline cranelift-codegen/src/simple_preopt.rs]: ditto

File: cranelift-codegen/src/simple_preopt.rs [hunk 13]
```diff
@@ -536,11 +542,245 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+
+    info.args.as_mut_slice(&mut pos.func.dfg.value_lists)[0] = info.cmp_arg;
+    if let InstructionData::Branch { ref mut opcode, .. } = pos.func.dfg[info.br_inst] {
+        *opcode = info.new_opcode;
+    } else {
+        panic!();
+    }
+}
+
+struct BranchOrderInfo {
```
Comment (bnjbvr) [inline cranelift-codegen/src/simple_preopt.rs]: Since this struct is only used internally within a function, could we use a tuple in the function instead, and that would remove the need for this structure?

File: cranelift-codegen/src/simple_preopt.rs [hunk 14]
```diff
@@ -536,11 +542,245 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+            if let Some(prev_dest) = prev_inst_data.branch_destination() {
+                if prev_dest != next_ebb {
+                    return;
+                }
+            } else {
+                return;
+            }
+
+            match prev_inst_data {
+                InstructionData::Branch {
```
Comment (bnjbvr) [inline cranelift-codegen/src/simple_preopt.rs]: Do we want to handle BranchInt (for `brif)` and BranchFloat (for `brff`) here too? (we could do this in a separate PR)

**Comment by bjorn3** [comment]
I believe you meant `other` instead of `order` in the pr title :)

File: cranelift-codegen/src/simple_preopt.rs [hunk 15]
```diff
@@ -536,11 +542,245 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+}
+
+/// Fold comparisons into branch operations when possible.
+///
+/// This matches against operations which compare against zero, then use the
+/// result in a `brz` or `brnz` branch. It folds those two operations into a
+/// single `brz` or `brnz`.
+fn branch_opt(pos: &mut FuncCursor, inst: Inst) {
+    let mut info = if let InstructionData::Branch {
+        opcode, ref args, ..
```
[the remark this answers is not in the document]
Comment (tyler) [reply cranelift-codegen/src/simple_preopt.rs]: Good point. Done.

File: cranelift-codegen/src/simple_preopt.rs [hunk 16]
```diff
@@ -536,11 +542,245 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
...
+
+        if let InstructionData::IntCompareImm {
+            opcode: Opcode::IcmpImm,
+            arg: cmp_arg,
+            cond: cmp_cond,
+            imm: cmp_imm,
+        } = pos.func.dfg[iconst_inst]
+        {
+            let cmp_imm: i64 = cmp_imm.into();
+            if cmp_imm != 0 {
```
[the remark this answers is not in the document]
Comment (tyler) [reply cranelift-codegen/src/simple_preopt.rs]: Doing this would require changing the definition of `Imm64(i64)` to `Imm64(pub i64)`, as far as I can tell. I could certainly do that, but figured there might be a reason why it's private.

I'd love to do that as `Imm64::from(0)` but https://github.com/rust-lang/rust/issues/57240 is not yet a thing.

File: cranelift-codegen/src/simple_preopt.rs (hunk 11 above)
Comment (tyler) [reply cranelift-codegen/src/simple_preopt.rs]: The trouble is that just below this I borrow `pos.func.dfg` as mutable. I haven't been able to figure out how to appease the borrow checker in this case. I'm all ears if you have suggestions though.

File: cranelift-codegen/src/simple_preopt.rs (hunk 13 above)
Comment (tyler) [reply cranelift-codegen/src/simple_preopt.rs]: I went ahead and did this. But I'm really not sure that it improved readability of the function. Take a look and let me know what you think?

File: cranelift-codegen/src/simple_preopt.rs (hunk 14 above)
Comment (tyler) [reply cranelift-codegen/src/simple_preopt.rs]: I don't *think* those should be necessary here, because they're not supposed to be constructed by frontends and this is a preopt pass. Please correct me if I'm thinking about that incorrectly though.

**Comment by tyler** [comment]
> I believe you meant `other` instead of `order` in the pr title :)

Surprisingly, I didn't. I was referring to re-ordering branches. :)

File: cranelift-codegen/src/simple_preopt.rs (hunk 11 above)
Comment (bnjbvr) [reply cranelift-codegen/src/simple_preopt.rs]: Sounds fair, I asked just in case.

File: cranelift-codegen/src/simple_preopt.rs (hunk 14 above)
Comment (bnjbvr) [reply cranelift-codegen/src/simple_preopt.rs]: Good point! I asked this since I didn't know these opcodes were private.

File: cranelift-codegen/src/simple_preopt.rs (hunk 13 above)
Comment (bnjbvr) [reply cranelift-codegen/src/simple_preopt.rs]: Definitely an improvement *after* the let statement. It's maybe less clear within the body, but it seems there are few enough lines in between that it's easy to check the tuple definition, so I'd be good with keeping it the new way. Thanks!

**Comment by bnjbvr** [comment]
Thanks, looks great!
